src/libstd/thread/local.rs

```diff
@@ -293,13 +294,14 @@ mod imp {
             Key {
                 inner: UnsafeCell::new(None),
                 dtor_registered: Cell::new(false),
-                dtor_running: Cell::new(false)
             }
         }
 
         pub unsafe fn get(&'static self) -> Option<&'static UnsafeCell<Option<T>>> {
-            if intrinsics::needs_drop::<T>() && self.dtor_running.get() {
-                return None
+            if intrinsics::needs_drop::<T>() {
+                if DTOR_RUNNING.with(|running| { running.get() }) {
```

Comment:
Unfortunately this change may be a little too-backwards-incompatible to work out well. I worry that this is "fixing" the problem by papering over some other underlying cause. One example of the breakage is the travis logs indicate a failing test which otherwise passes here.